sound/soc/sof/Makefile

```diff
@@ -3,7 +3,7 @@
 ccflags-y += -DDEBUG
```

Comment:
It's not exactly correct to say that the audio client platform's probe deals with loading the topology. The topology is loaded as follows: audio client register machine driver device -> which in turn calls machine driver probe -> which registers the sound card -> then when platform is bound to the card's dai link the platform's pcm_probe is called which loads the topology.

Reply:
@ranj063 The commit message should really explain the real need for this separation and also an ascii art picture with the arhitecture. 

Is this commit trying to model the real hardware arhitecture? On arm64 Device Tree is used to basically describe the hardware hierarchy as a tree with devices attached on buses.


In non-SOF usecases for a simple Audio scenario we have DAIs and codecs both corresponding to a piece of hardware (Digital Audio Interface IP and Codec IP). Both of these are standalone nodes in the Device Tree description.

Next: the link between these two is described inside the Device Tree by a "virtual" node which is basically the machine driver.

Now: In the case of SOF we have:

- the main platform device which is probed via ACPI / DT / PCI descriptors. Then according to your new approach this main platform device can have multiple clients. Each client may have some children.

Then in the particular case of Audio client it may have:
- machine driver
- component driver

This hierarchy looks not really natural to me, but we can work with it because it's really the old hierarchy with an additional node introduced for supporting multiple clients.

The reason I want to better understand this is that on the arm side we will use Device Tree files to describe this hierarchy (although we can syntetically created it like in the Intel side by register platform devices).

Reply:
I would like to see MFD generic driver layer being separated from SOF & Audio specifics since the idea is to support multiple MFD child devices, where some of them are non-audio related and some of the audio may be implemented based on the different FW ABI interface.

That is why we should avoid sof_mfd naming and choose something more generic (i.e. dsp_mfd). Additionally the /drivers/mfd path seems much more suitable location then sof folder path.

The main MFD device should focus on:
	* HW initialization
	* FW load
	* Providing API services (IPC client) to child devices (audio & other)
	* Determine what mfd child devices to create (depending on platform capabilities)
	
The SOF Audio Platform driver should be loaded on MFD child device and utilize MFD generic device API to communicate with the HW.

Since the MFD device design will have significant impact on the Audio future in Linux then I believe we should set up a Skype meeting to discuss this topic.

Reply:
@mwasko agree with the skype meeting. Also, I think @ranj063 doesn't want to use the MFD API just the naming is unfortunate.

Reply:
@lgirdwood While talking only about audio functionality with multiple cards / platforms per DSP / PCI device, it might be ok to use specialised APIs, but if we really want to be able to add non-audio functions in the future, I'd still consider making this a proper drivers/mfd driver.

Reply:
> @mwasko MFD core
> 
> > I would like to see MFD generic driver layer being separated from SOF & Audio specifics since the idea is to support multiple MFD child devices, where some of them are non-audio related and some of the audio may be implemented based on the different FW ABI interface.
> 
> "MFD" API here will only provide a "pipe" for upper client IPC (which can be any dialect), but the low level driver will stay in sound directory (since we are an audio HW device and audio is the main function).
> 
The DSP HW has become multi-function HW and should be treated as such, so the low level driver imho should be separated from audio path as this is only one of it functions (even if the so called main).
> > That is why we should avoid sof_mfd naming and choose something more generic (i.e. dsp_mfd). Additionally the /drivers/mfd path seems much more suitable location then sof folder path.
> 
> We are MFD in name only, and this just means more orchestartion for client developers with having to send patches to more development trees (and pull from more trees).
The final design of MFD generic driver should make it separate and allow to do changes within, while not requiring linked modifications in client MFD drivers. The architecture, once the client interface will be defined, should allow to develop low level MFD driver and client MFD function drivers independently.
> 
> > The main MFD device should focus on:
> > 
> > * HW initialization
> > * FW load
> 
> "base" FW load will be done by lower level automatically, but clients could certainly use IPCs to load modules.
> 
> > * Providing API services (IPC client) to child devices (audio & other)
> > * Determine what mfd child devices to create (depending on platform capabilities)
> > 
> > The SOF Audio Platform driver should be loaded on MFD child device and utilize MFD generic device API to communicate with the HW.
> > Since the MFD device design will have significant impact on the Audio future in Linux then I believe we should set up a Skype meeting to discuss this topic.
> 
> Lets adds this to topic for ELCE audio conf.
Sadly I won't be able to be present on the audio conf, Cezary will be. Anyway I hope we could have a sync meeting on this topic next week (before the conf take place)